### DISCUSSION TITLE
Issue: Segmentation Fault in multipath_basic_test()

### DISCUSSION THREAD

**Discussion by arpan21020** [private-octopus/picoquic#1788] [Q&A]
<h2>Description</h2>
<p>I am testing the multipath_test code and attempting to run the multipath_basic_test() function. However, it results in a segmentation fault. Upon debugging, I identified that within multipath_basic_test(), the function multipath_test_one() is called. This function invokes tls_api_init_ctx_ex2(), which is defined in tls_api_test.c.

To investigate further, I added print statements inside the definition of tls_api_init_ctx_ex2(). These revealed that while the client is successfully created, the picoquic_create() function returns NULL when attempting to create the server.</p>
<p><b>Could someone help identify why picoquic_create() is failing for the server context?</b></p>

**Comment by huitema** [comment]
The creation of a server side context will fail if the key and certificate are not properly created. The test keys are located in the source code. By default, the test assumes that it is running in the same directory as the source code. If it is not, you can use the "-S" option to picoquic_ct to document the location of the source code. If you fail to do that, the code will not find the test file, and will refuse to create the context.

That said, the code should not crash. Looks like a test bug.

**Reply by arpan21020** [reply]
I checked the code and in tls_api_test.c
there are two variables
char test_server_cert_file[512]="../ca-cert.pem";
char test_server_key_file[512]="../server-key.pem";

I am using the key that I generated while running the sample code of picoquic, do I have to generate the keys again?  Because it is still not working,
also there is one more variable that I have no idea
char test_server_cert_store_file[512];

Can you provide me with the command for generating the keys and certificates??

**Reply by huitema** [reply]
Most people use openssl to generate keys and certs. 

The tests are programmed to operate in a variety of environments. They use commands like:
```
ret = picoquic_get_input_path(test_server_cert_file, sizeof(test_server_cert_file), picoquic_solution_dir, PICOQUIC_TEST_FILE_SERVER_CERT);
```
to concatenate the path to the source files (picoquic_solution_dir) and the path to the file inside the solution (PICOQUIC_TEST_FILE_SERVER_CERT in that example). I suspect that when you are running your tests, the variable `picoquic_solution_dir` is not set correctly. What is the location (pwd) at which you are running `picoquic_ct`, and what is the path to the `picoquic` sources?

**Reply by arpan21020** [reply]
Thanks for your response. But I don't understand the use of picoquic_solution_dir, I set it to some random directory for instance to check.

I want to test the multipath functionality of the picoquic.
Can you guide me from scratch , how should I start, what all inputs I need to provide?
One thing I noticed that there is no main() function in the multipath_test.c, so I created one and called a function multipath_basic_test() which resulted in segmentation fault.
Am I doing something wrong here?

My aim is to stream video files using picoquic protocol. Please advise me how to proceed?

**Comment by huitema** [comment]
You don't need to build a main program. All the test functions can be called from `picoquic_ct` (`picoquic_t.exe` on Windows). Do something like:
```
picoquic_ct -S source_dir <name-of-test>
```
For example, if you built the code in a `build` folder of depot, and you want to run the basic multipath test you can do:
```
picoquic_ct -S .. multipath_basic
```

**Comment by huitema** [comment]
On the specific question of "stream video files using picoquic protocol", I suppose that you mean "stream video files using QUIC". There are 2 IETF working groups studying that, and proposing two different approaches. One is to use the real time transfer protcol, but instead of sending RTP packets over UDP, send them over a QUIC connection using QUIC datagrams -- see https://www.ietf.org/archive/id/draft-ietf-avtcore-rtp-over-quic-04.html. The other approach, "Media over QUIC" (MoQ), proposes sending media streams as series of QUIC streams and possibly datagrams, see https://datatracker.ietf.org/doc/draft-ietf-moq-transport/.

Either approach could be implemented using picoquic, using picoquic functions to send and receive datagrams or streams. Neither of these approaches requires the use of multipath. Use of multipath to transfer multiple media streams is indeed still a research project.

**Reply by arpan21020** [reply]
Thank you so much for your detailed response and the clarification about running test functions through `picoquic_ct`. That cleared up a lot of my confusion.  


I am also currently working on this research project focused on streaming media using multipath QUIC, and I aim to leverage picoquic for this. However, I am facing some challenges in understanding the multipath functionality in picoquic and how to make custom modifications to `multipath_test.c` for my specific use case.  

Could you guide me on how to approach modifying and testing the `multipath_test.c` file? Specifically:  
1. How can I build and run a modified version of `multipath_test.c` for testing purposes?  
2. Are there any additional inputs or configurations I need to consider when using multipath for media streaming?  

Currently, when I runned the rest I am getting this error
![image](https://github.com/user-attachments/assets/685de359-5880-4203-8d88-73052357bb0d)

Your insights would be incredibly helpful for me to proceed. Thanks again for your time!

**Comment by huitema** [comment]
As the name indicates, `multipath_test.c` is test code, meant to run in the test environment. You can use it to get some idea on how to use the various APIs, but it is not designed to run outside the test environment.

Do you have a version of your scenario running in a single path environment, i.e., without trying to create new paths? That might be a nice first step.

**Reply by arpan21020** [reply]
Could you please guide me on how to properly build multipath_test.c after making modifications? I’ve added some print statements in the file, but the changes don’t seem to reflect even after running the `make` command again.

Additionally, I encountered an error earlier (shared in the previous image). Could you help me understand and resolve it?

**Reply by huitema** [reply]
There is no way to build multipath_test.c in isolation. This is not a supported scenario. It can only be built as part of picoquic_t.

**Comment by arpan21020** [comment]
![image](https://github.com/user-attachments/assets/af20f0c0-3227-4acb-8544-8b96aa849dfc)
<h2>Can you help me out with this error</h2>

This is my solution dir, it has the certificate and the key
![image](https://github.com/user-attachments/assets/634e5490-0258-4dd6-9f65-15d80dd6abbc)

**Comment by huitema** [comment]
That's an errant message. The internal API `picoquic_find_path_by_address` is used to match incoming packets to the appropriate "path" context within a connection. In that case, if there is no path, we have an error. The same API is also used to check inside the `picoquic_probe_new_path_ex` API to check that the proposed path does not already exist. This is what we are seeing here, and "Could notfind path by addr" is the expected result, not an error.

But yes, that's confusing. There should not be a debug message in the "normal" case. It should be fixed.

The test actually succeeds. Two paths are created, and the load is shared between the two paths.

**Reply by arpan21020** [reply]
![image](https://github.com/user-attachments/assets/86fb147f-e070-41d6-92c1-ef89fca0f713)
![image](https://github.com/user-attachments/assets/ec4aaecb-12f0-4227-812a-dbf4eb4466d6)
Could you provide a brief overview of the `test_scenario_multipath`, including how data transmission occurs and what type of data is exchanged between the client and server during multipath testing?

**Reply by huitema** [reply]
The test programs uses a simulation to execute scenarios. Each scenario line describes a bidirectional stream and has 4 components:

* the stream-id over which the exchange executes
* the id of the stream after which the stream starts executing. For example, if the second line in the example said "{ 8, 4, ...}" instead of "{8, 0, ...}, the second stream would start only after all data on the first stream (id=4) has been exchanged.
* the number of bytes sent by the client.
* the number of bytes sent by the server.

As these are test programs, the bytes exchanged are arbitrary. The tests will verify that the bytes received are equal to the bytes sent.

The multipath tests, by default, open 2 paths between client and server and use both paths to send data. There are a variety of tests, for both basic and path failure tests. All the tests have a target "simulated execution time", to verify that the data is sent efficiently.

**Comment by huitema** [comment]
As for the general approach, my advice remains to first implement a single path application, test it, and after that add code to create additional paths.
